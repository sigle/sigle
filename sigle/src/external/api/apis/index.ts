/* tslint:disable */
/* eslint-disable */
export * from './AnalyticsApi';
export * from './SubscriptionApi';
export * from './UserApi';
