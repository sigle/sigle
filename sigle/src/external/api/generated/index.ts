/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
export { ApiError } from './core/ApiError';
export { CancelablePromise, CancelError } from './core/CancelablePromise';
export { OpenAPI } from './core/OpenAPI';
export type { OpenAPIConfig } from './core/OpenAPI';

export type { AddEmailDto } from './models/AddEmailDto';
export type { AnalyticsHistoricalDto } from './models/AnalyticsHistoricalDto';
export type { AnalyticsStoryDto } from './models/AnalyticsStoryDto';
export type { ContactsListsEntity } from './models/ContactsListsEntity';
export type { CreateSubscriberDto } from './models/CreateSubscriberDto';
export type { CreateUserFollowDto } from './models/CreateUserFollowDto';
export type { DeleteUserFollowDto } from './models/DeleteUserFollowDto';
export type { EmailDto } from './models/EmailDto';
export type { ExploreResponse } from './models/ExploreResponse';
export type { ExploreUser } from './models/ExploreUser';
export type { HistoricalDto } from './models/HistoricalDto';
export type { NewsletterEntity } from './models/NewsletterEntity';
export type { PublicNewsletterEntity } from './models/PublicNewsletterEntity';
export type { PublishStoryDto } from './models/PublishStoryDto';
export type { ReferrerDto } from './models/ReferrerDto';
export type { SenderEntity } from './models/SenderEntity';
export type { SendTestStoryDto } from './models/SendTestStoryDto';
export type { StoryDto } from './models/StoryDto';
export type { SubscriptionDto } from './models/SubscriptionDto';
export type { UnpublishStoryDto } from './models/UnpublishStoryDto';
export type { UpdateContactsListDto } from './models/UpdateContactsListDto';
export type { UpdateNewsletterDto } from './models/UpdateNewsletterDto';
export type { UpdateSenderDto } from './models/UpdateSenderDto';
export type { UserMeProfileEntity } from './models/UserMeProfileEntity';
export type { UserProfileEntity } from './models/UserProfileEntity';
export type { VerifyEmailDto } from './models/VerifyEmailDto';

export { AnalyticsService } from './services/AnalyticsService';
export { DefaultService } from './services/DefaultService';
export { NewslettersService } from './services/NewslettersService';
export { StoriesService } from './services/StoriesService';
export { SubscribersService } from './services/SubscribersService';
export { SubscriptionService } from './services/SubscriptionService';
export { UserService } from './services/UserService';
