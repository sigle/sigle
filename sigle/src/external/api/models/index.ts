/* tslint:disable */
/* eslint-disable */
export * from './ApiAnalyticsHistoricalGet200Response';
export * from './ApiAnalyticsHistoricalGet200ResponseHistoricalInner';
export * from './ApiAnalyticsHistoricalGet200ResponseStoriesInner';
export * from './ApiAnalyticsReferrersGet200ResponseInner';
export * from './ApiSubscriptionsCreatorPlusPost200Response';
export * from './ApiSubscriptionsCreatorPlusPostRequest';
export * from './ApiSubscriptionsGet200Response';
export * from './ApiUsersMeGet200Response';
