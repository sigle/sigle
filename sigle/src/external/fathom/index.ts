import { aggregate } from './aggregate';
import { aggregatePath } from './aggregatePath';
import { aggregateReferrers } from './aggregateReferrers';

export interface FathomClientParams {
  apiToken: string;
  entityId: string;
}

export const initFathomClient = (clientParams: FathomClientParams) => {
  return {
    aggregate: aggregate(clientParams),
    aggregatePath: aggregatePath(clientParams),
    aggregateReferrers: aggregateReferrers(clientParams),
  };
};

export const fathomClient = initFathomClient({
  apiToken: process.env.FATHOM_API_TOKEN!,
  entityId: process.env.FATHOM_ENTITY_ID!,
});
