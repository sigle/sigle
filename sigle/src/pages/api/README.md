# Sigle API

## Endpoints

### Stats historical data

`https://app.sigle.io/api/analytics/historical` - Get the historical statistics.

Sample Response for days grouping:

```json
[
  {
    "date": "2022-04-29",
    "visits": 0,
    "pageviews": 0
  },
  {
    "date": "2022-04-30",
    "visits": 0,
    "pageviews": 0
  }
  // ...
]
```

Sample Response for months grouping:

```json
[
  {
    "date": "2022-03",
    "visits": 0,
    "pageviews": 0
  },
  {
    "date": "2022-04",
    "visits": 0,
    "pageviews": 0
  }
  // ...
]
```

#### Parameters

- **`dateFrom`** - The date from which to get the statistics (e.g. 2022-04-01).
- **`dateGrouping`** - The date grouping (e.g. day, month). When day is set the date format is YYYY-MM-DD. When month is set the date format is YYYY-MM.
- `storyId` - optional - The story id to get the statistics for.
