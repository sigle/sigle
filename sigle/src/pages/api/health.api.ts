import { NextApiHandler } from 'next';

/**
 * Health check endpoint, used to check that the server is running properly
 */
const healthCheckEndpoint: NextApiHandler = async (_, res) => {
  res.status(200).send(true);
};

export default healthCheckEndpoint;
