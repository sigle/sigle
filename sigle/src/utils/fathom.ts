export enum Goals {
  LOGIN = 'IHBJQCC0',
  CREATE_NEW_STORY = 'WYESYDDJ',
  PUBLISH = 'PYKUM6ZG',
  FREE_USERNAME_CREATED = 'OR1RD9A7',
}
