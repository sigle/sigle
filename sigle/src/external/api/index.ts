import { sigleConfig } from '../../config';
import { OpenAPI } from './generated';

OpenAPI.BASE = sigleConfig.apiUrl!;
OpenAPI.WITH_CREDENTIALS = true;

export * from './generated';
