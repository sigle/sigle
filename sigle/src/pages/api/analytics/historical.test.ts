/* eslint-disable @typescript-eslint/no-explicit-any */
import { analyticsHistoricalEndpoint } from './historical.api';
import { fathomClient } from '../../../external/fathom';

jest.mock('../../../external/fathom');

it('Should throw an error if dateFrom is missing', async () => {
  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: {} } as any,
    {
      status,
    } as any
  );
  expect(status).toBeCalledWith(400);
  expect(json).toBeCalledWith({ error: 'dateFrom is required' });
});

it('Should throw an error if dateFrom is invalid', async () => {
  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: 'invalid' } } as any,
    {
      status,
    } as any
  );
  expect(status).toBeCalledWith(400);
  expect(json).toBeCalledWith({ error: 'dateFrom is invalid' });
});

it('Should throw an error if dateGrouping is missing', async () => {
  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: '2022-05-04' } } as any,
    {
      status,
    } as any
  );
  expect(status).toBeCalledWith(400);
  expect(json).toBeCalledWith({ error: 'dateGrouping is required' });
});

it('Should throw an error if dateGrouping is invalid', async () => {
  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: '2022-05-04', dateGrouping: 'invalid' } } as any,
    {
      status,
    } as any
  );
  expect(status).toBeCalledWith(400);
  expect(json).toBeCalledWith({ error: 'dateGrouping must be day or month' });
});

it('Should throw an error if dateGrouping is day and dateFrom > 2 month', async () => {
  jest
    .useFakeTimers()
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: '2022-01-04', dateGrouping: 'day' } } as any,
    {
      status,
    } as any
  );
  expect(status).toBeCalledWith(400);
  expect(json).toBeCalledWith({
    error: 'dateFrom must be within 2 months when grouping by days',
  });
});

it('Respond with a formatted time series for days', async () => {
  // Set a fake timer so we can verify the end date
  jest
    .useFakeTimers()
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  (fathomClient.aggregatePath as jest.Mock).mockResolvedValue([]);

  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: '2022-03-02', dateGrouping: 'day' } } as any,
    {
      status,
    } as any
  );
  expect(fathomClient.aggregatePath).toBeCalledTimes(25);
  expect(status).toBeCalledWith(200);
  expect(json).toMatchSnapshot();

  (fathomClient.aggregatePath as jest.Mock).mockReset();
});

it('Respond with a formatted time series for months', async () => {
  // Set a fake timer so we can verify the end date
  jest
    .useFakeTimers()
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  (fathomClient.aggregatePath as jest.Mock).mockResolvedValue([]);

  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    { query: { dateFrom: '2022-01-02', dateGrouping: 'month' } } as any,
    {
      status,
    } as any
  );
  expect(fathomClient.aggregatePath).toBeCalledTimes(25);
  expect(status).toBeCalledWith(200);
  expect(json).toMatchSnapshot();

  (fathomClient.aggregatePath as jest.Mock).mockReset();
});

it('Respond with a formatted time series for one story', async () => {
  // Set a fake timer so we can verify the end date
  jest
    .useFakeTimers()
    .setSystemTime(new Date('2022-04-04 12:00:00 UTC').getTime());
  (fathomClient.aggregatePath as jest.Mock).mockResolvedValue([]);

  const json = jest.fn();
  const status = jest.fn(() => ({
    json,
  }));

  await analyticsHistoricalEndpoint(
    {
      query: { dateFrom: '2022-03-02', dateGrouping: 'day', storyId: 'test' },
    } as any,
    {
      status,
    } as any
  );
  expect(fathomClient.aggregatePath).toBeCalledTimes(1);
  expect(status).toBeCalledWith(200);
  expect(json).toMatchSnapshot();
});
