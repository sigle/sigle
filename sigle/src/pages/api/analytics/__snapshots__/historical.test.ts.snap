// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Respond with a formatted time series for days 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "date": "2022-03-02",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-03",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-04",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-05",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-06",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-07",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-08",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-09",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-10",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-11",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-12",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-13",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-14",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-15",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-16",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-17",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-18",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-19",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-20",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-21",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-22",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-23",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-24",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-25",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-26",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-27",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-28",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-29",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-30",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-31",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-01",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-02",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-03",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-04",
          "pageviews": 0,
          "visits": 0,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Respond with a formatted time series for months 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "date": "2022-01",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-02",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04",
          "pageviews": 0,
          "visits": 0,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Respond with a formatted time series for one story 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "date": "2022-03-02",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-03",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-04",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-05",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-06",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-07",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-08",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-09",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-10",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-11",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-12",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-13",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-14",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-15",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-16",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-17",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-18",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-19",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-20",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-21",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-22",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-23",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-24",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-25",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-26",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-27",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-28",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-29",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-30",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-03-31",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-01",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-02",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-03",
          "pageviews": 0,
          "visits": 0,
        },
        Object {
          "date": "2022-04-04",
          "pageviews": 0,
          "visits": 0,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
